megabus/src/main/java/com/bazaarvoice/megabus/resolver/MegabusRefResolver.java

```diff
@@ -147,39 +146,53 @@ private ResolutionResult resolveRefs(Iterator<MegabusRef> refs) {
             } catch (UnknownTableException | UnknownPlacementException e) {
                 // take no action and discard the event, as the table was deleted before we could process the event
                 _discardedMeter.mark();
+            } catch (Throwable t) {
+                _errorProcessingCounter.inc();
+                throw t;
             }
         }
 
-        Iterator<DataProvider.AnnotatedContent> readResultIter = annotatedGet.execute();
+        try {
+            Iterator<DataProvider.AnnotatedContent> readResultIter = annotatedGet.execute();
 
-        List<Map<String, Object>> resolvedDocuments = new ArrayList<>();
-        List<MegabusRef> missingRefs = new ArrayList<>();
+            List<Map<String, Object>> resolvedDocuments = new ArrayList<>();
+            List<MegabusRef> missingRefs = new ArrayList<>();
 
-        readResultIter.forEachRemaining(result -> {
-            Map<String, Object> content = result.getContent();
-            AtomicBoolean triggerEvent = new AtomicBoolean(false);
+            readResultIter.forEachRemaining(result -> {
+                try {
+                    Map<String, Object> content = result.getContent();
+                    AtomicBoolean triggerEvent = new AtomicBoolean(false);
 
-            refTable.row(Coordinate.fromJson(content)).forEach((changeId, ref) -> {
+                    refTable.row(Coordinate.fromJson(content)).forEach((changeId, ref) -> {
 
-                if (result.isChangeDeltaPending(changeId)) {
-                    _pendingMeter.mark();
-                    missingRefs.add(ref);
-                    return;
-                }
+                        if (result.isChangeDeltaPending(changeId)) {
+                            _pendingMeter.mark();
+                            missingRefs.add(ref);
+                            return;
+                        }
 
-                if (result.isChangeDeltaRedundant(changeId)) {
-                    _redundantMeter.mark();
-                    return;
-                }
+                        if (result.isChangeDeltaRedundant(changeId)) {
+                            _redundantMeter.mark();
+                            return;
+                        }
 
-                triggerEvent.set(true);
+                        triggerEvent.set(true);
+                    });
+
+                    if (triggerEvent.get()) {
+                        resolvedDocuments.add(content);
+                    }
+                } catch (Throwable t) {
+                    _errorProcessingCounter.inc();
```

Comment:
Won't an error thrown in this closure be counted twice by metrics because of the nested catch?

Reply:
oooh, yeah  I guess from the re-throw, heh. I was trying to avoid declaring empty variables outside the local scope because I wanted to catch any exception from Emo from `annotatedGet.execute()`.

Originally I was counting the documents and incrementing the counter by the number, in an attempt to count the number of "documents" in error, but I  realized any error is gonna stop the thread when it throws an exception, so it might as  well just be "one" document, no matter where it is in the batch  (and if there are multiple, then it's the "first" one, and the others will be counted later).

Reply:
actually because of all of this, I'm going to go a different route and just wrap the call to `resolveRefs` in a try/catch instead.